Fix code format scripts compat with non-GNU Unices

akien-mga opened the pull request:
And improve it to also remove trailing tabs, not just spaces.

Not tested on macOS myself but at least this is documented as working on https://stackoverflow.com/a/15416836
CC @RevoluPowered or @bruvzg to confirm.

@aaronfranke The script takes a while to format the codebase, it would be nice if we can find some optimizations:
```
$ time misc/scripts/file_format.sh
real    1m8,844s
user    0m56,079s
sys     0m12,323s
```

akien-mga committed:
Style: Fix code format scripts compat with non-GNU Unices

It's too hard to get compatibility between GNU and BSD sed,
so let's just use perl oneliners.

And improve it to also remove trailing tabs, not just spaces.

Calinou commented:
> The script takes a while to format the codebase, it would be nice if we can find some optimizations:

Maybe it could be run in parallel if `parallel` is available in the PATH.

bruvzg commented:
```
stat: illegal option -- c
usage: stat [-FlLnqrsx] [-f format] [-t timefmt] [file ...]
Files in this commit comply with the formatting rules.
```

And it is creating `filename-e` files for every singe file in the repo.

bruvzg commented:
GNU `stat` is available in homebrew `coreutils` package as `gstat`.

aaronfranke commented:
> Alternatively, the tool can be modified to format only modified files (according to Git).

I don't think there is really an issue with the CI running it on all files in PRs. Plus if there is a problem I think it's best to "be noisy" about it so that it gets fixed fast. The Static Checks action doesn't really take that long overall, so if we can make it faster with something like `parallel` that would be nice I guess, but it's not really that slow IMO. On my system:

```
$ time misc/scripts/file_format.sh 
real	0m36.572s
user	0m29.118s
sys	0m8.100s
```

By the way, on the topic of "`parallel` isn't available by default" - well you already need `dos2unix` and `recode`, one more dependency doesn't hurt I suppose.

bruvzg commented on misc/scripts/file_format.sh:
```diff
@@ -30,13 +30,13 @@ while IFS= read -rd '' f; do
     dos2unix "$f" 2> /dev/null
     # Ensures that files do not contain a BOM.
-    sed -i '1s/^\xEF\xBB\xBF//' "$f"
+    sed -i -e '1s/^\xEF\xBB\xBF//' "$f"
```
>-i extension
>             Edit files in-place, saving backups with the specified extension.  If a zero-length extension is given, no backup will be saved.  It is not recommended to give a zero-length extension when in-place editing files, as you risk corruption or partial content in situations where disk space is exhausted, etc.

This is what causing `-e` files, adding "" after `-i` seems to fix it.

```suggestion
    sed -i "" -e '1s/^\xEF\xBB\xBF//' "$f"
```
but it won't work with GNU sed.

akien-mga commented:
I replaced `sed` usage by `perl -pi -e` and dropped usage on GNU `stat`.

@bruvzg @RevoluPowered Can you test if it now works on macOS?

Performance wise, I found that half the time is spent in `recode`, so maybe we can optimize there a bit by finding a faster way to ensure encoding.
The two perl one liners take ~10 to ~15 s each. It's possible that the `dos2unix` call (~8 s) is no longer needed if `perl -pli -e` handles fixing line endings already (didn't check).

bruvzg commented:
> Can you test if it now works on macOS?

Seems to work fine.

> and dropped usage on GNU stat.

`stat` is also used in the `black_format.sh` and `clang_format.sh` scripts.

bruvzg commented:
And maybe it should check if `dos2unix` and `recode` are installed and print error message (neither is installed on macOS be default, `perl` is).

akien-mga commented:
> `stat` is also used in the `black_format.sh` and `clang_format.sh` scripts.

Fixed.

> And maybe it should check if `dos2unix` and `recode` are installed and print error message (neither is installed on macOS be default, `perl` is).

Done.

akien-mga commented:
Cherry-picked for 3.2.3.

aaronfranke commented on misc/scripts/file_format.sh:
```diff
@@ -29,22 +35,17 @@ while IFS= read -rd '' f; do
     # Remove trailing space characters.
-    sed -z -i 's/\x20\x0A/\x0A/g' "$f"
+    # -l option handles newlines conveniently and seems to also get rid of BOMs.
+    perl -i -ple 's/\s*$//g' "$f"
```
I just tested this locally and it doesn't get rid of BOMs. Here's a test file which contains a BOM, after running `perl -i -ple 's/\s*$//g' test.txt`, the file still contains a BOM. [test.txt](https://github.com/godotengine/godot/files/4986397/test.txt)

akien-mga replied:
Well the script removed BOMs for me. I assumed it was this change, but maybe it's recode or dos2unix that does it.

bruvzg commented:
`misc/scripts/black_format.sh` is still broken.

```
Formatting Python files...
usage: find [-H | -L | -P] [-EXdsx] [-f path] path ... [expression]
       find [-H | -L | -P] [-EXdsx] -f path [path ...] [expression]

No Path provided. Nothing to do 😴
```

`find` seems to be different too, as it was with the previous issues, changing it to the GNU version (`gfind` form `brew install findutils`) works fine.
